### PR TITLE
Update README and update build script for vcpkg

### LINKED ISSUE #89: if build.sh is run twice, it errors out
first it failed because ninja wasn't installed, so i called the build.sh again, resulting in:

```
CMake Error at cmake/vcpkg_helper.cmake:17 (message):
  Please define a path to VCPKG_ROOT.  See
  https://github.com/trailofbits/cxx-common for more details.  Or if you
  don't want to use vcpkg dependencies, add '-DUSE_SYSTEM_DEPENDENCIES=ON'
Call Stack (most recent call first):
  CMakeLists.txt:22 (include)
```

i then rm -rf'd the rellic-build dir, downloaded the 500 MB of stuff again, just to get
```
  Could NOT find Git (missing: GIT_EXECUTABLE)
```
(i assumed if i cloned the repo before entering the ubuntu rootfs i wouldnt need it)

running build.sh again results in the same error as above.
i guess i have to wait another hour to download the 500MB again and see what comes before the next error.

### PR DESCRIPTION
**Pull request by ekilmer**
Try to fix issues brought up in #89 

* Does not try to redownload the pre-built library tar ball

closes #89

### CODE REVIEW

**Comment by rofl0r** [comment]
thanks, trying this branch, i get:

```
./rellic/scripts/build.sh
[-] Library version is vcpkg_ubuntu20.04_llvm-9_amd64
/root/lifting-bits-downloads /root/rellic-build
Fetching: https://github.com/trailofbits/cxx-common/releases/latest/download/vcpkg_ubuntu20.04_llvm-9_amd64.tar.xz and placing in /root/lifting-bits-downloads
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   170  100   170    0     0    492      0 --:--:-- --:--:-- --:--:--   492
100     9  100     9    0     0     16      0 --:--:-- --:--:-- --:--:--    16
+ tar -xJf vcpkg_ubuntu20.04_llvm-9_amd64.tar.xz --warning=no-timestamp
xz: (stdin): File format not recognized
tar: Child returned status 1
tar: Error is not recoverable: exiting now
[x] Unable to download vcpkg libraries build vcpkg_ubuntu20.04_llvm-9_amd64.
[x] Build aborted.


# file lifting-bits-downloads/vcpkg_ubuntu20.04_llvm-9_amd64.tar.xz
lifting-bits-downloads/vcpkg_ubuntu20.04_llvm-9_amd64.tar.xz: ASCII text, with no line terminators
```

contents of that file: `Not Found`

**Comment by ekilmer** [comment]
Can you try to install the following
```
apt update && apt install pixz xz-utils
```
and then re-run the code?

**Comment by rofl0r** [comment]
alright. i renamed my ninja binary (actually it is samurai) to ninja.bk and repulled this branch 990ba13 and installed the apt stuff as you mentioned.

```
# rellic/scripts/build.sh 
[-] Library version is libraries-llvm900-ubuntu20.04-amd64
[-] Z3 version is z3-4.7.1-x64-ubuntu-16.04
Fetching: https://github.com/trailofbits/cxx-common/releases/latest/download/libraries-llvm900-ubuntu20.04-amd64.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   175  100   175    0     0    504      0 --:--:-- --:--:-- --:--:--   505
100   666  100   666    0     0   1202      0 --:--:-- --:--:-- --:--:--  650k
100  296M  100  296M    0     0  4032k      0  0:01:15  0:01:15 --:--:-- 6004k
[+] Z3 already downloaded, skipping
[+] Configuring...
CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
-- Configuring incomplete, errors occurred!
See also "/root/rellic-build/CMakeFiles/CMakeOutput.log".
[x] Build aborted.
```
the download succeeds now, as the typo s/9/900/ is fixed.

```
# mv /usr/bin/ninja-1 /usr/bin/ninja
# rellic/scripts/build.sh 
[-] Library version is libraries-llvm900-ubuntu20.04-amd64
[-] Z3 version is z3-4.7.1-x64-ubuntu-16.04
[+] Configuring...
-- Using new vcpkg build system
CMake Error at cmake/vcpkg_helper.cmake:17 (message):
  Please define a path to VCPKG_ROOT.  See
  https://github.com/trailofbits/cxx-common for more details.  Or if you
  don't want to use vcpkg dependencies, add '-DUSE_SYSTEM_DEPENDENCIES=ON'
Call Stack (most recent call first):
  CMakeLists.txt:22 (include)


-- Configuring incomplete, errors occurred!
See also "/root/rellic-build/CMakeFiles/CMakeOutput.log".
[x] Build aborted.
```

**Comment by ekilmer** [comment]
Also, @rofl0r sorry for the confusion, but the initial changes I made to `build.sh` already exist as `build_with_vcpkg.sh`. I updated the README to make this more clear and hopefully prevent more people from getting stuck with the old build system that we're phasing out.

What you'll want to do is follow the README in this PR. I've tested with a base Docker image, which I think should be close to what you'll have as well

**Comment by rofl0r** [comment]
is there actually a fix committed that picks up the tarballs if they already exist and passes the same options to cmake on a second run ?

**Comment by ekilmer** [comment]
> is there actually a fix committed that picks up the tarballs if they already exist and passes the same options to cmake on a second run ?

I believe adding the `-z` option as shown here https://github.com/lifting-bits/rellic/pull/91/files#diff-8ef37c104e75c1df347d4ecc1c6c8fa75f335855cb0f7fc13c0eb075fedf1ee9R110-R117 (lines 110-117 in the build_with_vcpkg.sh script) should prevent redownloads of the tar (unless the tar is newer online)

If you can, please use the `./scripts/build_with_vcpkg.sh` script

**Comment by rofl0r** [comment]
alright, following the new instructions and removed git binary (i removed it to test whether the functionality i complained about now works), i get as expected

```
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Z3: /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libz3.a (found suitable version "4.8.9", minimum required is "4.7.1") 
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Git (missing: GIT_EXECUTABLE)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindGit.cmake:96 (find_package_handle_standard_args)
  /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/scripts/buildsystems/vcpkg.cmake:622 (_find_package)
  cmake/git_watcher.cmake:75 (find_package)
  rellic/Version/CMakeLists.txt:11 (include)


-- Configuring incomplete, errors occurred!
See also "/root/rellic/rellic-build/CMakeFiles/CMakeOutput.log".
See also "/root/rellic/rellic-build/CMakeFiles/CMakeError.log".
[x] Build aborted.
```
checking out 9dff942 ... and entering the chroot again and running the build command again (i actually forgot to make git available again):
it now works:
```
/root/rellic# ./scripts/build_with_vcpkg.sh --llvm-version 10
...
-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libLLVMDebugInfoCodeView.a /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libLLVMDebugInfoMSF.a /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libLLVMOption.a /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libLLVMSupport.a -lrt -ldl -lm /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libLLVMDemangle.a /root/lifting-bits-downloads/vcpkg_ubuntu-20.04_llvm-10_amd64/installed/x64-linux-rel/lib/libz3.a -lpthread ../rellic/Version/libRellicVersion.a 
make[2]: Leaving directory '/root/rellic/rellic-build'
[100%] Built target rellic-headergen-10.0
make[2]: Leaving directory '/root/rellic/rellic-build'
[100%] Built target rellic-decomp-10.0
make[1]: Leaving directory '/root/rellic/rellic-build'
/usr/bin/cmake -E cmake_progress_start /root/rellic/rellic-build/CMakeFiles 0
```
good job, thanks. is the .0 fix you just pushed critical ?

**Comment by ekilmer** [comment]
> it now works:

Woohoo!! 🎉 Great!!

> is the .0 fix you just pushed critical ?

No, that was just an update to the README, in case someone wants to follow it _exactly_.

I believe https://github.com/lifting-bits/rellic/commit/9dff942253c1f3166e95858935ae3882a8fbb2cc was the last commit to the build script 👍

**Comment by rofl0r** [comment]
> is the .0 fix you just pushed critical ?

ah i see, it was just for docs. unfortunately the issue #90 is still not fixed with the timezone install:

```
./rellic-build/tools/rellic-decomp-10.0 --input /root/ags272/agsedit.exe.bc --output agsedit.c
error: unknown target triple '', please use -triple or -arch
*** Aborted at 1611625399 (unix time) try "date -d @1611625399" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGSEGV (@0x0) received by PID 1519 (TID 0x7f7302f84ec0) from PID 0; stack trace: ***
    @     0x7f73034dd3c0 (/home/rofl0r/ub/root/usr/lib/x86_64-linux-gnu/libpthread-2.31.so+0x153bf)
    @     0x500001d9e1e6 (/home/rofl0r/ub/root/root/rellic/rellic-build/tools/rellic-decomp-10.0+0x1d9e1e5)
    @     0x5000008dea2b (/home/rofl0r/ub/root/root/rellic/rellic-build/tools/rellic-decomp-10.0+0x8dea2a)
Segmentation fault
```

**Comment by ekilmer** [comment]
Thank you for being so patient! 

The directions in the README very easily go out of date, but we will try our best to keep them updated.

Let me know if you have any other questions 🙂

**Comment by ekilmer** [comment]
> unfortunately the issue #90 is still not fixed with the timezone install

Hmmmm. I might have to bring in @surovic to take a look at that error.

Would you be able to share the file to help reproduce it? If not that's fine, but it might be faster to debug on our end if we have that as a way to reproduce the error

**Comment by rofl0r** [comment]
certainly, here it is: https://0x0.st/--eM.xz - it's the result of running retdec 4.0 on https://github.com/ags-archives/engines/blob/master/ags272/agsedit.exe - though i think as i mentioned in #90 that the date error is due to some supposedly installed, yet not found ubuntu component.

**Comment by ekilmer** [comment]
Just playing around with this, I get the following in `gdb`:
```
root@590f235976b6:/work# gdb --args ./rellic-build/tools/rellic-decomp-10.0 --input output-90 --output /dev/stderr
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./rellic-build/tools/rellic-decomp-10.0...
(gdb) r
Starting program: /work/rellic-build/tools/rellic-decomp-10.0 --input output-90 --output /dev/stderr
warning: Error disabling address space randomization: Operation not permitted
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
error: unknown target triple '', please use -triple or -arch

Program received signal SIGSEGV, Segmentation fault.
0x000055a51551b1e6 in clang::CompilerInstance::createPreprocessor(clang::TranslationUnitKind) ()
(gdb) bt
#0  0x000055a51551b1e6 in clang::CompilerInstance::createPreprocessor(clang::TranslationUnitKind) ()
#1  0x000055a51405ba2b in rellic::InitCompilerInstance (ins=..., target_triple="") at /work/rellic/AST/Util.cpp:55
#2  0x000055a513ef8f4f in (anonymous namespace)::GeneratePseudocode (output=..., module=...) at /usr/include/c++/9/bits/basic_string.h:936
#3  main (argc=<optimized out>, argv=<optimized out>) at /work/tools/decomp/Decomp.cpp:202
(gdb)
```

Where the `target_triple` at `#1` is a zero-length string. That triple comes from here https://github.com/lifting-bits/rellic/blob/50dff6eb07b70ec9c34321af6e50a2b4523b7675/tools/decomp/Decomp.cpp#L63 so I think the bitcode might be malformed?

I disassembled the bitcode file you uploaded (thank you!) with `llvm-dis-10 -o issue-90.ll issue-90.bc` and I don't see a triple listing. Whereas when I use clang to compile the bitcode (as shown in the README), I get a line like the following 
```
target triple = "x86_64-pc-linux-gnu"
```

Might be worth playing around with inserting a line like that?

**Comment by ekilmer** [comment]
Okay after adding the following line near the top of the disassembled `.ll` file (after `target datalayout`)
```
target triple = "x86_64-pc-win32-unknown"
```
(not sure if that's the correct triple...) and reassembling (`llvm-as-10 issue-90.ll -o issue-90-fix.bc`) I get the following error:

```
root@590f235976b6:/work# ./rellic-build/tools/rellic-decomp-10.0 --input issue-90-fix.bc --output /dev/stderr
F0125 21:34:33.084846  8653 IRToASTVisitor.cpp:127] Check failed: !result.isNull() Unknown LLVM Type
*** Check failure stack trace: ***
    @     0x55865dd1984c  google::LogMessageFatal::~LogMessageFatal()
    @     0x55865c81e9d6  _ZN6rellic14IRToASTVisitor11GetQualTypeEPN4llvm4TypeE.localalias
Aborted
```

Definitely need to have @surovic take a look at this.

**Comment by rofl0r** [comment]
thanks for your analysis! i think the triple should be `i486-pc-win32-unknown`, since it's a 32bit binary. i will try to follow the steps you mentioned tomorrow as well. maybe i can figure out where the date error is coming from with strace or gdb.

**Comment by surovic** [comment]
> ```
> root@590f235976b6:/work# ./rellic-build/tools/rellic-decomp-10.0 --input issue-90-fix.bc --output /dev/stderr
> F0125 21:34:33.084846  8653 IRToASTVisitor.cpp:127] Check failed: !result.isNull() Unknown LLVM Type
> *** Check failure stack trace: ***
>     @     0x55865dd1984c  google::LogMessageFatal::~LogMessageFatal()
>     @     0x55865c81e9d6  _ZN6rellic14IRToASTVisitor11GetQualTypeEPN4llvm4TypeE.localalias
> Aborted
> ```
> 
> Definitely need to have @surovic take a look at this.

So there's apparently a `i864` type inside the bitcode, which points to something retdec specific happening. Seems like the following bitcode, or something similar is the cause.
```llvm
define i32 @function_5ba465() local_unnamed_addr {
dec_label_pc_5ba465:
  %0 = call i32 @__asm_wait(), !insn.addr !336912
  %1 = call i32 @__asm_fnsave(), !insn.addr !336913
  %2 = sext i32 %1 to i864, !insn.addr !336913
  %3 = call i32 @function_5ba85c(i864 %2, i32 ptrtoint (i32* @199 to i32), i32 ptrtoint (i32* @199 to i32), i32 ptrtoint (i32* @199 to i32), i32 ptrtoint (i32* @199 to i32)), !insn.addr !336914
  call void @__asm_frstor(i864 %2), !insn.addr !336915
  ret i32 %3, !insn.addr !336916
}
```
It's an interesting issue, but given the extreme specificity of this, I'll have to ask @rofl0r to submit a PR if he wants this fixed soon.

File: README.md [hunk 1]
```diff
@@ -51,33 +51,38 @@ sudo apt-get install \
...
+If the distribution you're on doesn't include a recent release of CMake (3.14 or later), you'll need to install it. For Ubuntu, see here https://apt.kitware.com/.
+
 The next step is to clone the Rellic repository.
 
 ```shell
 git clone https://github.com/trailofbits/rellic.git
 ```
 
-Finally, we build Rellic. This script will create another directory, `rellic-build`, in the current working directory. All remaining dependencies needed by Rellic will be built in the `rellic-build` directory. Rellic will also download Z3 in an appropriate version if it's not installed in the system.
+Finally, we build Rellic. This script will create another directory, `rellic-build`, in the current working directory. All remaining dependencies needed by Rellic will be download and place in the parent directory alongside the repo checkout in `lifting-bits-downloads` (see the script's `-h` option for more details).
```
Comment (surovic) [inline README.md:72]: > download**ed** and place**d**
typos

**Review by surovic — APPROVED** [review]

**Pull request merged**